### PR TITLE
Carousel: Make controlled component

### LINKED ISSUE #1958: Carousel needs the capacity to be a controlled component
Currently carousel is lacking certain features that allow it to be a controlled component. While more edge-case, this limits applications from handling user interactions with the carousel itself. The following features should be added:

* `currentPanel` should be a prop that controls the active panel, falling back to state only if undefined.
* `onRequestPanelChange` should be a function that fires when the right / left navigation arrows or indicator dots are hit that passes the currentPanel and requestedPanel (the panel that would be switched to based on user interaction) as data.
* `isAutoplayOn` should be a prop that controls the autoplay feature, falling back to state only if undefined.
* `onRequestAutoplayToggle` should be a function that fires whenever the pause/play button is hit passing the current isAutoplayOn value as data
* The `indicatorStyles` prop currently does nothing and should be removed

### PR DESCRIPTION
**Pull request by kevinparkerson**
Fixes #1958

### NOTE: MERGE #1973 FIRST

Just a few notes with this one:
- I changed all "AutoPlay" instances to "Autoplay" to be consistent with the `hasAutoplay` prop. The assistive text changed to `autoplayButton` when doing this, otherwise nothing else changed API-wise. Capitalizing the P in autoplay would have changed much more of the API (though this is a prototype component so does it matter?)
- Should it be `onRequestPanelChange` or `onPanelChange`? Same for `onRequestAutoplayToggle`. I wasn't sure but for some reason `onRequestPanelChange` feels more correct
- I didn't even realize `isInfinite` was an available prop that was supposed to control whether the carousel can loop back to the first or last panel while navigating. It wasn't working properly, but I have now fixed it so it behaves at it should. Additionally, I made autoplay either stop or loop depending on what the value of that prop is. Would you agree that is appropriate behavior?
- I removed the `indicatorStyles` prop as it did nothing and seemed pointless

---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [x] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [x] `npm run lint:fix` has been run and linting passes.
* [x] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [x] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [x] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [x] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [x] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] TravisCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

**Comment by interactivellama** [comment]
Here is the lowdown on callback names https://github.com/salesforce/design-system-react/blob/master/docs/codebase-overview.md#prefix-onrequestaction-to-pre-state-change-callbacks-functions

File: components/carousel/__docs__/__snapshots__/storybook-stories.storyshot [hunk 1]
```diff
@@ -63,7 +63,7 @@ exports[`DOM snapshots SLDSCarousel 3 items NoImageTest 1`] = `
             className="slds-carousel__panels slds-is-relative"
             style={
               Object {
-                "transform": "translateX(0px)",
+                "transform": "translateX(NaNpx)",
```
Comment (interactivellama) [inline components/carousel/__docs__/__snapshots__/storybook-stories.storyshot]: `NaNpx` seems concerning.

File: components/carousel/index.jsx [hunk 2]
```diff
@@ -78,7 +78,11 @@ class Carousel extends React.Component {
 			PropTypes.string,
 		]),
 		/**
-		 * Boolean showing whether the autoplay feature is available or not
+		 * Dictates the currently active/visible carousel panel. 1-indexed. Use with `onRequestPanelChange` for a controlled carousel component. If not provided, the carousel will manage this itself via state.
```
Comment (interactivellama) [inline components/carousel/index.jsx]: I'm not sure I know of anything else 1-indexed in the library.

File: components/carousel/__docs__/__snapshots__/storybook-stories.storyshot (hunk 1 above)
Comment (kevinparkerson) [reply components/carousel/__docs__/__snapshots__/storybook-stories.storyshot]: I believe these are the result of code within canUseDOM / canUseEventListeners conditionals not running. There are actually 56 instances of this in the snapshot, and each one is linked to numbers determined by the stageWidth DOM measurement.

File: components/carousel/index.jsx (hunk 2 above)
Comment (kevinparkerson) [reply components/carousel/index.jsx]: Changed to 0 indexing & pushing soon

**Comment by interactivellama** [comment]
Merged! Do you think you could write a jQuery version of this Carousel next week?

**Comment by interactivellama** [comment]
🤣

**Pull request merged**